File: modules/appmodules/servicito/src/main/java/com/example/servicito/domains/charges/models/pojo/Charge.java
package com.example.servicito.domains.charges.models.pojo;

import com.fasterxml.jackson.annotation.JsonIgnore;

import javax.persistence.Column;
import javax.persistence.Embeddable;
import javax.validation.constraints.NotNull;

@Embeddable
public class Charge {
    private int gasCharge;
    private int waterCharge;
    private int electricityCharge;
    private int securityCharge;
    @NotNull
    @Column(nullable = false)
    private int rent;
    private int advance;

    @JsonIgnore
    public int getTotalUtilityCharges() {
        int charge = 0;
        if (getWaterCharge() > 0) charge += getWaterCharge();
        if (getGasCharge() > 0) charge += getGasCharge();
        if (getElectricityCharge() > 0) charge += getElectricityCharge();
        if (getSecurityCharge() > 0) charge += getSecurityCharge();
        return charge;
    }

    @JsonIgnore
    public int getTotalCharges() {
        return this.getRent() + this.getTotalUtilityCharges();
    }

    @JsonIgnore
    public String getWaterChargeStr() {
        if (this.waterCharge < 0) return "*ব্যাবহারের উপর নির্ভরশীল*";
        if (this.waterCharge == 0) return "*ভাড়ার অন্তর্ভূক্ত*";
        return this.waterCharge + " টাকা";
    }

    @JsonIgnore
    public String getGasChargeStr() {
        if (this.gasCharge < 0) return "*ব্যাবহারের উপর নির্ভরশীল*";
        if (this.gasCharge == 0) return "*ভাড়ার অন্তর্ভূক্ত*";
        return this.gasCharge + " টাকা";
    }

    @JsonIgnore
    public String getElectricityChargeStr() {
        if (this.electricityCharge < 0) return "*ব্যাবহারের উপর নির্ভরশীল*";
        if (this.electricityCharge == 0) return "*ভাড়ার অন্তর্ভূক্ত*";
        return this.electricityCharge + " টাকা";
    }

    @JsonIgnore
    public String getSecurityChargeStr() {
        if (this.securityCharge < 0) return "*ব্যাবহারের উপর নির্ভরশীল*";
        if (this.securityCharge == 0) return "*ভাড়ার অন্তর্ভূক্ত*";
        return this.securityCharge + " টাকা";
    }


    public int getGasCharge() {
        return gasCharge;
    }

    public void setGasCharge(int gasCharge) {
        this.gasCharge = gasCharge;
    }

    public int getWaterCharge() {
        return waterCharge;
    }

    public void setWaterCharge(int waterCharge) {
        this.waterCharge = waterCharge;
    }

    public int getElectricityCharge() {
        return electricityCharge;
    }

    public void setElectricityCharge(int electricityCharge) {
        this.electricityCharge = electricityCharge;
    }

    public int getSecurityCharge() {
        return securityCharge;
    }

    public void setSecurityCharge(int securityCharge) {
        this.securityCharge = securityCharge;
    }

    public int getRent() {
        return rent;
    }

    public void setRent(int rent) {
        this.rent = rent;
    }

    public int getAdvance() {
        return advance;
    }

    public void setAdvance(int advance) {
        this.advance = advance;
    }
}

File: modules/appmodules/servicito/src/main/kotlin/com/example/servicito/domains/apartments/models/embedables/Charge.kt
package com.example.servicito.domains.apartments.models.embedables

import javax.persistence.Column
import javax.persistence.Embeddable

@Embeddable
class Charge {
    @Column(name = "gas_charge")
    var gasCharge = 0

    @Column(name = "water_charge")
    var waterCharge = 0

    @Column(name = "electricity_charge")
    var electricityCharge = 0

    @Column(name = "security_charge")
    var securityCharge = 0

    @Column(name = "rent", nullable = false)
    var rent = 0

    @Column(name = "advance")
    var advance = 0


    fun calculateTotalUtilityCharges(): Int {
        var charge = 0
        if (waterCharge > 0) charge += waterCharge
        if (gasCharge > 0) charge += gasCharge
        if (electricityCharge > 0) charge += electricityCharge
        if (securityCharge > 0) charge += securityCharge
        return charge
    }

    fun totalCharges(): Int {
        return rent + calculateTotalUtilityCharges()
    }

    fun getWaterChargeStr(): String {
        if (waterCharge < 0) return "*ব্যাবহারের উপর নির্ভরশীল*"
        return if (waterCharge == 0) "*ভাড়ার অন্তর্ভূক্ত*" else "$waterCharge টাকা"
    }

    fun getGasChargeStr(): String {
        if (gasCharge < 0) return "*ব্যাবহারের উপর নির্ভরশীল*"
        return if (gasCharge == 0) "*ভাড়ার অন্তর্ভূক্ত*" else gasCharge.toString() + " টাকা"
    }

    fun getElectricityChargeStr(): String {
        if (electricityCharge < 0) return "*ব্যাবহারের উপর নির্ভরশীল*"
        return if (electricityCharge == 0) "*ভাড়ার অন্তর্ভূক্ত*" else electricityCharge.toString() + " টাকা"
    }

    fun getSecurityChargeStr(): String {
        if (securityCharge < 0) return "*ব্যাবহারের উপর নির্ভরশীল*"
        return if (securityCharge == 0) "*ভাড়ার অন্তর্ভূক্ত*" else securityCharge.toString() + " টাকা"
    }

}